pkg/export/otel/traces.go

```diff
@@ -137,39 +145,125 @@ func traceAttributes(span *transform.HTTPRequestSpan) []attribute.KeyValue {
+			semconv.NetPeerPort(span.HostPort),
+			semconv.HTTPRequestContentLength(int(span.ContentLength)),
+		}
 	}
 	return []attribute.KeyValue{}
 }
 
+func traceName(span *transform.HTTPRequestSpan) string {
+	switch span.Type {
+	case transform.EventTypeHTTP:
+		name := span.Method
+		if span.Route != "" {
+			name += " " + span.Route
+		}
+		return name
+	case transform.EventTypeGRPC:
+		return span.Path
+	case transform.EventTypeHTTPClient:
+		return span.Method
+	}
+	return ""
+}
+
+func spanKind(span *transform.HTTPRequestSpan) trace2.SpanKind {
+	switch span.Type {
+	case transform.EventTypeHTTP, transform.EventTypeGRPC:
+		return trace2.SpanKindServer
+	case transform.EventTypeHTTPClient:
+		return trace2.SpanKindClient
+	}
+	return trace2.SpanKindInternal
+}
+
+func makeSpan(parentCtx context.Context, tracer trace2.Tracer, span *transform.HTTPRequestSpan) SessionSpan {
+	var spP trace2.Span
+	// Create a parent span for the whole request session
+	ctx, sp := tracer.Start(parentCtx, traceName(span),
+		trace2.WithTimestamp(span.RequestStart),
+		trace2.WithSpanKind(spanKind(span)),
+		trace2.WithAttributes(traceAttributes(span)...),
+	)
+
+	if span.RequestStart != span.Start {
+		// Create a child span showing the queue time
+		_, spQ := tracer.Start(ctx, "in queue",
+			trace2.WithTimestamp(span.RequestStart),
+			trace2.WithSpanKind(trace2.SpanKindInternal),
+		)
+		spQ.End(trace2.WithTimestamp(span.Start))
+
+		// Create a child span showing the processing time
+		// Override the active context for the span to be the processing span
+		ctx, spP = tracer.Start(ctx, "processing",
+			trace2.WithTimestamp(span.Start),
+			trace2.WithSpanKind(trace2.SpanKindInternal),
+		)
+		spP.End(trace2.WithTimestamp(span.End))
+	}
+
+	sp.End(trace2.WithTimestamp(span.End))
+
+	return SessionSpan{*span, ctx}
+}
+
 func (r *TracesReporter) reportTraces(input <-chan []transform.HTTPRequestSpan) {
 	defer r.close()
 	tracer := r.traceProvider.Tracer(reporterName)
 	for spans := range input {
 		for i := range spans {
-			attrs := traceAttributes(&spans[i])
-
-			// Create a parent span for the whole request session
-			ctx, sp := tracer.Start(context.TODO(), "session",
-				trace2.WithTimestamp(spans[i].RequestStart),
-				trace2.WithAttributes(attrs...),
-				trace2.WithSpanKind(trace2.SpanKindServer),
-			)
-
-			// Create a child span showing the queue time
-			_, spQ := tracer.Start(ctx, "in queue",
-				trace2.WithTimestamp(spans[i].RequestStart),
-				trace2.WithSpanKind(trace2.SpanKindInternal),
-			)
-			spQ.End(trace2.WithTimestamp(spans[i].Start))
-
-			// Create a child span showing the processing time
-			_, spP := tracer.Start(ctx, "processing",
-				trace2.WithTimestamp(spans[i].Start),
-				trace2.WithSpanKind(trace2.SpanKindInternal),
-			)
-			spP.End(trace2.WithTimestamp(spans[i].End))
-
-			sp.End(trace2.WithTimestamp(spans[i].End))
+			span := &spans[i]
+
+			if span.Type == transform.EventTypeHTTPClient {
```

Comment:
since `gocyclo` threshold has been decreased to 10, maybe it complains about this method. In that case I'd suggest to split some blocks into methods/functions.